File: DSA_python/DSA02033 SỐ XA CÁCH.py
def submit():
    n = int(input())
    store = [0]*n
    placed = [False]*(n+1)
    
    def recursion(store, placed, j):
        if j == n:
            print("".join(map(str, store)))
            return
        
        for i in range(1, n+1):
            if placed[i] == False:

                if j > 0 and abs(store[j-1]-i) == 1: continue
                store[j] = i
                placed[i] = True

                recursion(store, placed, j+1)
                placed[i] = False


    recursion(store, placed, 0)
    print()


test = int(input())
for i in range(test):
    submit()

File: DSA/DSA02033 SỐ XA CÁCH.cpp
#include <bits/stdc++.h>
using namespace std;

int n;

void recursion(vector<int> store, vector<bool> &placed, int j)
{
    if (j == n)
    {
        for (auto &&k : store)
        {
            cout << k;
        }
        cout << endl;
        return;
    }

    for (int i = 1; i <= n; i++)
    {
        if (!placed[i])
        {
            if (j > 0 && abs(store[j-1]-i) == 1) continue;

            store[j] = i;
            placed[i] = true;
            recursion(store, placed, j+1);
            placed[i] = false;
        }
    }
}

void submit()
{
    cin >> n;
    vector<int> store(n);
    vector<bool> placed(n+1, false);
    
    recursion(store, placed, 0);
    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int test;
    cin >> test;
    while (test--)
    {
        submit();
    }
    

    return 0;
}